SAK-34741 SAKAI-2735 SAM-1113 Ability to flag question as Extra Credit

davidpbauer opened the pull request:
This allows instructors to flag questions in their assessment as extra credit. Extra credit quesitons
will not count towards the total possible points for an assessment. However, any points earned by the
student for answering the extra credit question correctly will be added to their total earned points
for that assessment. Currently, students are able to score over 100 percent on an assessment.

davidpbauer commented:
If I remember correctly we added two new columns (one for working copy and one for published) to track the extra credit setting.
`
alter table sam_publisheditem_t add ISEXTRACREDIT number(1);
alter table sam_item_t add ISEXTRACREDIT number(1);
`

bjones86 commented:
@davidpbauer, if this feature requires database conversion, you'll have to submit a PR to add the snippets to the conversion scripts:

https://github.com/sakaiproject/sakai-reference/tree/master/docs/conversion

bjones86 commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentBean.java:
```diff
@@ -179,7 +179,9 @@ public void setQuestionSizeAndTotalScore() {
           if (item.getItemData().getScore()!=null){
-            this.totalScore += item.getItemData().getScore().doubleValue();
+            if(item.getItemData().getIsExtraCredit()==null || !item.getItemData().getIsExtraCredit().booleanValue()) {
+              this.totalScore += item.getItemData().getScore().doubleValue();
```
I think calling `.booleanValue()` and `.doubleValue()` are unnecessary here.

bjones86 commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/ItemBean.java:
```diff
@@ -102,6 +102,7 @@
   
   private int totalMCAsnwers;
   private CalculatedQuestionBean calculatedQuestion;
+  private boolean isExtraCredit;
```
Lombok this to avoid writing boilerplate code below.

bjones86 commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/PublishedAssessmentBeanie.java:
```diff
@@ -136,7 +136,8 @@ public void setQuestionSizeAndTotalScore() {
       for (ItemContentsBean item : items) {
         if (item.getItemData().getScore()!=null){
-          this.totalScore += item.getItemData().getScore().doubleValue();
+          if(item.getItemData().getIsExtraCredit() == null || !item.getItemData().getIsExtraCredit().booleanValue())
```
Same deal here with `booleanValue()` and `doubleValue()`

bjones86 commented on samigo/samigo-app/src/webapp/jsf/delivery/deliverAssessment.jsp:
```diff
@@ -327,6 +327,7 @@ document.links[newindex].onclick();
 					<f:convertNumber maxFractionDigits="2" groupingUsed="false"/>
 				</h:outputText>
 			</span>
+            <h:outputText styleClass="extraCreditLabel" rendered="#{question.itemData.isExtraCredit == true}" value="#{deliveryMessages.extra_credit_preview}" />
```
Indentation is a little off here.

bjones86 commented on samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/assessment/ItemData.java:
```diff
@@ -77,7 +77,8 @@
   private String hash;
   private Long originalItemId;
- 
+  private Boolean isExtraCredit;
```
Lombok

bjones86 commented on samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/assessment/PublishedAssessmentData.java:
```diff
@@ -703,7 +703,9 @@ public Double getTotalScore(){
       while (iter2.hasNext()){
         PublishedItemData item = (PublishedItemData)iter2.next();
-        total = total.add(BigDecimal.valueOf(item.getScore().doubleValue()));
+        if( item.getIsExtraCredit() == null || !item.getIsExtraCredit().booleanValue()) {
```
`booleanValue()` and `doubleValue()`

bjones86 commented on samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/assessment/PublishedItemData.java:
```diff
@@ -73,6 +73,7 @@
   private Date createdDate;
   private String lastModifiedBy;
   private Date lastModifiedDate;
+  private Boolean isExtraCredit;
```
Lombok

bjones86 commented on samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/PublishedAssessmentFacade.java:
```diff
@@ -745,7 +745,9 @@ public Double getTotalScore(){
       while (iter2.hasNext()){
         ItemDataIfc item = (ItemDataIfc)iter2.next();
-        total = total.add(BigDecimal.valueOf(item.getScore().doubleValue()));
+        if (item.getIsExtraCredit()==null || !item.getIsExtraCredit().booleanValue()) {
```
`booleanValue()` and `doubleValue()`

davidpbauer commented:
> @davidpbauer, if this feature requires database conversion, you'll have to submit a PR to add the snippets to the conversion scripts (MySQL and Oracle):
> 
> https://github.com/sakaiproject/sakai-reference/tree/master/docs/conversion

I've made a pull request for the new extra credit column https://github.com/sakaiproject/sakai-reference/pull/71

bjones86 commented:
Great, thanks @davidpbauer.

davidpbauer commented:
Thanks, @bjones86. I've fixed the indent, removed the unboxing, and switched to using lombok.

@ern is the datatype something you want me to address on this pull request? When I originally wrote this I was following the pattern already used. I'm happy to update it, but I need to jump back to other work so I'm not sure I would have the time to change and test it for a little while.

ern commented:
I am fine with the code as you mentioned it's an existing pattern, but wanted to leave it up to you.

bjones86 commented:
@davidpbauer, we have a small conflict here, sorry! I'll merge this in after the conflict is resolved.

davidpbauer commented:
I resolved the conflict

bjones86 commented:
Thanks! I'll merge as soon as Travis is done the build.
